OrbitLinuxTracing/ContextSwitchAndThreadStateVisitor.cpp

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ContextSwitchAndThreadStateVisitor.h"
+
+#include <absl/container/flat_hash_map.h>
+
+#include <algorithm>
+#include <utility>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitLinuxTracing/TracerListener.h"
+
+namespace LinuxTracing {
+
+using orbit_grpc_protos::SchedulingSlice;
+using orbit_grpc_protos::ThreadStateSlice;
+
+void ContextSwitchAndThreadStateVisitor::ProcessInitialTidToPidAssociation(pid_t tid, pid_t pid) {
```

Comment:
This seems kind of flaky. we are reading system-wide mapping from /proc but it is not atomic so we have mostly correct tid->pid association. Is there a way to make it more correct?

Reply:
(maybe we should have a process in OrbitService for monitoring FORKs all the time on order to have a mapping always on so to speak, in which case we do not need to do mostly correct initialization)

Reply:
It is and we are aware of it, that's why first enable the `perf_event_open` file descriptors and then call `RetrieveTidToPidAssociationSystemWide()`. As this seems to be fine, always monitoring forks is way overkill.

Reply:
>  why we first enable the perf_event_open file descriptors

That is, we get all fork events already before we read from /proc/? Sounds like that would work correctly or are there also cases where that can fail?

Reply:
> that initial state when OrbitService starts has to be correct

Good point.

> What are the user facing effects of getting this wrong? Some timeslices associated to the wrong process I think?

Exactly.

> Should we PTRACE_INTERRUPT/CONT to make sure state can't change during initialization?

That sounds intrusive and way more than we need, we can make this work without it.

> That is, we get all fork events already before we read from /proc/? Sounds like that would work correctly or are there also cases where that can fail?

The idea is that if we miss a new process or thread while reading from `/proc`, then we should have caught the `PERF_RECORD_FORK`. 
Thinking about it, the one corner case might be if a thread exits before reading from `/proc`. We might have context switches related to that thread at the beginning of the capture without the process. 
Again, that would be at the very beginning of a capture, where things are never going to be all correct as some kind of non-zero-time initialization needs to happen. But if we really want to we should be able to record the association by using `PERF_RECORD_EXIT`.

Reply:
Ok, thanks for the clarification. 

> That sounds intrusive and way more than we need, we can make this work without it.

It's only once for a very brief moment at the start of the capture, but it requires to attach to the process, which will not work if a debugger is attached. So, yes, if we can make it work without it, that's very much preferred.

Reply:
I added a simple commit to also use `PERF_RECORD_EXIT`. It should cover the corner case I mentioned above. If you think it's too much I can just remove it, but if this now covers all cases (or bring us closer to it), then I'm happy about it. The frequency of these events is relatively low anyway.

Reply:
> But if we really want to we should be able to record the association by using PERF_RECORD_EXIT.

Well, that works only if the `PERF_RECORD_EXIT` comes before the context switch(es) with the -1 pid. I need to check that.

Reply:
Ok, I'll need to double check everything but that works, so I added a commit that also uses `PERF_RECORD_EXIT` to cover the corner case mentioned above. I can just remove it if we think it's too much (it covers the last scheduling slice of a thread that is exiting when the capture has just started, right before we have read from `/proc`), but the frequency of these events is generally low so it shouldn't hurt. 
(As a side note, if it was only for context switches this would be enough the not use `/proc` at all, but we still need a tid-to-pid association always up to date for thread state.)